src/Squidex.Infrastructure.MongoGridFs/Assets/MongoGridFsAssetStore.cs

```diff
@@ -0,0 +1,219 @@
+﻿// ==========================================================================
+//  Squidex Headless CMS
+// ==========================================================================
+//  Copyright (c) Squidex UG (haftungsbeschraenkt)
+//  All rights reserved. Licensed under the MIT license.
+// ==========================================================================
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Driver;
+using MongoDB.Driver.GridFS;
+
+namespace Squidex.Infrastructure.Assets
+{
+    public class MongoGridFsAssetStore : IAssetStore, IInitializable
+    {
+        private readonly string path;
+        private readonly IGridFSBucket bucket;
+        private readonly DirectoryInfo directory;
+
+        public MongoGridFsAssetStore(IGridFSBucket bucket, string path)
+        {
+            Guard.NotNull(bucket, nameof(bucket));
+            Guard.NotNullOrEmpty(path, nameof(path));
+
+            this.bucket = bucket;
+            this.path = path;
+
+            directory = new DirectoryInfo(path);
+        }
+
+        public void Initialize()
+        {
+            try
+            {
+                // test bucket
+                bucket.Database.ListCollections();
+
+                if (!directory.Exists)
+                {
+                    directory.Create();
+                }
+            }
+            catch (MongoException ex)
+            {
+                throw new ConfigurationException(
+                    $"Cannot connect to Mongo GridFS bucket '${bucket.Options.BucketName}'.", ex);
+            }
+            catch (IOException ex)
+            {
+                if (!directory.Exists)
+                {
+                    throw new ConfigurationException($"Cannot access directory '{directory.FullName}'", ex);
+                }
+            }
+        }
+
+        public string GenerateSourceUrl(string id, long version, string suffix)
+        {
+            var file = GetFile(id, version, suffix);
+
+            return file.FullName;
+        }
+
+        public async Task CopyAsync(string name, string id, long version, string suffix,
+            CancellationToken ct = default(CancellationToken))
+        {
+            try
+            {
+                var file = GetFile(name);
+
+                using (var stream = new MemoryStream())
+                {
+                    await bucket.DownloadToStreamByNameAsync(file.Name, stream, cancellationToken: ct);
```

Comment:
Can we not just rename the asset?

Reply:
no support from IGridFSBucket, unfortunately :(

Reply:
Or at least a version where you do not keep everything in memory.

Reply:
Rename != Copy

Reply:
no :) , copy means you have two copies at the end

Reply:
it would make sense to use Rename if you'll rename IAssetStore.CopyAsync to IAssetStore.RenameAsync

Reply:
Right

Reply:
Yes, as I said you are right. I was thinking too much about the use case here, where it does not really matter if it is copy or rename because the source will be deleted anyway ;)